### ISSUE TITLE
Using clustering for more than 2 classes of sentiments

### ISSUE THREAD

**Issue by flarrow7** [rafaljanwojcik/Unsupervised-Sentiment-Analysis#4]
Hi! I am trying to classify reviews in 4 different classes based on the degree of negative sentiment: Severely Negative, Negative, Less Negative, Rarely negative/positive. Do you have any suggestions using this approach?

**Comment by rafaljanwojcik** [comment]
Hey, thanks for reaching out! I guess it depends on how complicated are relationships in your data - if your dataset contains simple reviews as in case of the dataset I analyzed in my article, then you could try to create four clusters with KMeans (you can specify how many clusters there will be) instead of two, and see if your dataset can actually be split into meaningful clusters with this method. If not, you should probably try to label the data (most obvious solution) and train supervised model. I'm currently 'pondering' quite extensively on how this technique, or similar/more advanced to one I developed could be used for more complicated relationships in text data, but I didn't come to any conclusions yet :)

**Comment by flarrow7** [comment]
Thanks @rafaljanwojcik ! Yes, currently, apart from unsupervised learning I am looking into semi-supervised techniques like co-training and PU-learning. 
Also, I had a quick question, what if for a supervised we develop a Multi perception model or similar Deep neural network architecture model and try to obtain maximum validation accuracy. Now can this model be used for any similar unseen dataset? Cheers!

**Comment by rafaljanwojcik** [comment]
Hey, I have long and late answer for your quick question :)  I don't know the techniques you mentioned, but it sounds interesting :) What you're asking about is transfer learning - it really depends on the model you use, the data it is pertained on, and your downstream task data (the data you want to label). If the downstream task data is similar and can be thought of as coming from the same distribution as pertaining set, then your model could perform well. So the best option is to try 🙂 It should surely perform better than in case of randomly initialized model, and you could also try labeling only small fraction of training data, i.e., 1000 examples, and checking if the model already learns relatively good dependencies and performs well 🙂 

Current state-of-the-art pertained language models are Transformers - I really recommend you to try them out, i.e., BERT base fine-tuned on sentiment analysis, such models tend to perform really good, and are really easy to fine-tune to amazing performance with little amount of training data, thanks to their extensive pre-training. Link to the best repo in my opinion for working with the Transformers:
https://github.com/huggingface/transformers

If you're interested in how to fine-tune BERT on your task-specific data, I can also recommend my repository (a little self-advertisement ;) ):
https://github.com/rafaljanwojcik/SentenceBERT_vs_SiameseLSTM
I evaluated there two BERT-based models, one with classification layer architecture, and one with siamese architecture, on semantic similarity task. If you find it too hard to go through, or too task-specific, there are lots of tutorials online on how to use HuggingFace, i.e.,:
https://huggingface.co/transformers/training.html
